File: 79-word-search/79-word-search.cpp
class Solution {
public:
    
    bool dfs(int r, int c, vector<vector<char>>& board, vector<vector<int>>&vis, string &word, int idx){
        if(idx == word.size())
            return true;
       if(r >= board.size() || r<0 || c<0 || c>=board[0].size() || vis[r][c]==true ||
         word[idx]!=board[r][c])
           return false;
        if(word[idx] == board[r][c]){
            vis[r][c] = true;
        bool res = dfs(r+1, c, board, vis, word, idx+1)
            ||  dfs(r-1, c, board, vis, word, idx+1)
            ||  dfs(r, c+1, board, vis, word, idx+1)
            ||  dfs(r, c-1, board, vis, word, idx+1);
        vis[r][c] = false;
        return res;
        }
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
         vector<vector<int>>vis(board.size(), vector<int>(board[0].size(), 0));
        for(int i = 0;i<board.size(); i++){
            for(int j = 0; j<board[0].size(); j++){
                if(board[i][j]==word[0]){
                    if(dfs(i, j, board, vis, word, 0)){
                        return true;
                    }
                }
            }
        }
        
        return false;
    }
};

File: 79-word-search/79-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n = len(board)
        m = len(board[0])
        vis = set()
        
        def backtrack(i, j, idx):
            if len(word) == idx:
                return True
            
            if(i< 0 or i>=n or j<0 or j>=m or board[i][j]!=word[idx] or (i, j) in vis):
                return False
            
            if(board[i][j] == word[idx]):
                vis.add((i, j))
                res = (backtrack(i+1, j, idx+1) or
                    backtrack(i, j+1, idx+1) or
                    backtrack(i-1, j, idx+1) or
                    backtrack(i, j-1, idx+1))
                vis.remove((i, j))
                
            return res
        
        for i in range(n):
            for j in range(m):
                if word[0] == board[i][j]:
                    if backtrack(i, j, 0):
                        return True
                    
        return False